### DISCUSSION TITLE
Data ingestion protocol (TCP)

### DISCUSSION THREAD

**Discussion by collimarco** [zinclabs/zinc#127] [Q&A]
**Is it possible to use a plain TCP connection (with SSL) for data ingestion?** Basically keep a TCP connection open to continuously ship the logs.

I would like to understand how to ship logs directly from a Ruby on Rails application to Zinc, **without making an expensive HTTP request for each log line**. Currently we ship logs to Logz.io (ELK) with a simple TCP connection (not HTTP).

Thank you in advance

**Comment by prabhatsharma** [comment]
Currently, it is not possible to use TCP. I have been thinking about adding grpc support. Would that be helpful?

In general, I would recommend against writing logs directly from your RoR application to the logging system as it violates the [12 factor app logging approach](https://12factor.net/logs)

**Reply by collimarco** [reply]
Thanks for the reply! 

I don't know about grpc, it's the first time that I hear about it. I was hoping for something basic and simple, like a bare TCP connection where you simply write the log output to the socket (instead of a log file). This is probably the most basic implementation that can be easily implemented in any language (without any additional libraries or plugins). 

1. Is there any reason to use something more complex like grpc rather than a good old TCP connection?
2. Also, what protocol are you using for Fluent integration? Does it send HTTP requests to Zinc?

Thanks

**Reply by prabhatsharma** [reply]
grpc will be simpler as there is much better understanding about it and it is standardized. However in your case you should be using neither tcp or grpc. A better option would be to simply do a stdout.

e.g. puts("my logging text")

this stdout can then be picked up by a log forwarder like fluent bit and sent to whatever destination. This is the easiest way to do logging. If you are using docker then docker picks up these logs and stores them in a file in a rolling mode which can then be picked up by fluentbit. if you are running your code directly on a server then you can use fluentbit stdin mode - https://docs.fluentbit.io/manual/pipeline/inputs/standard-input

In your current mechanism you are violating the 12 factor app principle for logging.

**Reply by collimarco** [reply]
I cannot use the standard input solution (pipe), because the Rails app / server is started with Systemd. The 12 factor can be a good start, but sometimes there are reasons that make it useful to use different strategies... The 12 factor is not the Bible. We have been using this method of logging (TCP) with ELK for years and we never had any problem. 

What protocol are you using for **Fluent** integration? **Does it send HTTP requests to Zinc or what?**

Thanks

**Comment by prabhatsharma** [comment]
You don't really have to use a log forwarder like fluent bit if you don't want to. Not everyone uses it. It's just one of the things that I found to be working well for many companies (I have worked with 100s of AWS customers helping them build systems in my work at AWS).

Fluent bit uses HTTP for zinc. 

Next best option would be for you to call zinc HTTP APIs asynchronously from your app.

**Reply by collimarco** [reply]
Thanks. 
So if I understand correctly, Zinc at the moment ONLY supports the "single record" HTTP API and the "bulk" HTTP API: is that correct? 
And all the other integrations that you offer, like Fluent and Syslog-ng, are just built over that two methods?

**Reply by prabhatsharma** [reply]
100% correct. Fluentbit and syslog-ng both utilize bulk api
